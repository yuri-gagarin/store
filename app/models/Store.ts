import mongoose, { Schema, Document } from "mongoose";
import { IStorePicture } from "./StorePicture";
type IStorePictureRef = mongoose.Types.ObjectId | IStorePicture;

export interface IStore extends Document {
  title: string;
  description: string;
  images: IStorePictureRef[];
  createdAt: Date;
  editedAt?: Date;
}

const StoreSchema: Schema = new Schema({
  title: {
    type: String,
    required: true
  },
  description: {
    type: String,
    required: true
  },
  images: [
    {
      type: Schema.Types.ObjectId,
      ref: "StorePicture"
    }
  ],
  createdAt: {
    type: Date,
    default: Date.now,
    required: true
  },
  editedAt: {
    type: Date
  }
});
const Store = mongoose.model<IStore>('Store', StoreSchema);
export default Store;