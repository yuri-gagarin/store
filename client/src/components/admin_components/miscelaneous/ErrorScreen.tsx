import React from 'react'
import { Button, Dimmer, Header, Icon, Segment } from 'semantic-ui-react'
import axios from "axios";

type RetryLastReqFunc = <T> () => T
interface Props {
  errorMessage?: string;
  lastRequest?: () => Promise<void>;
}
const ErrorScreen: React.FC<Props> = ({ errorMessage, lastRequest }): JSX.Element => {

  const handleRetry = (e: React.MouseEvent<HTMLButtonElement>): void => {
    if (lastRequest) {
      lastRequest();
    }
  };
  
  return (
    <div style={{ height: "100vh" }}> 
      <Segment style={{ height: "100vh" }}>
        <Dimmer active>
          <Header as='h2' icon inverted>
            <Icon name='exclamation circle' />
              An Error Occured!
            <Header.Subheader>Dimmer sub-header</Header.Subheader>
            <Button 
              id="errorScreenRetryButton"
              inverted 
              color="green" 
              content="Retry" 
              onClick={handleRetry} 
            />
          </Header>
        </Dimmer>
      </Segment>
    </div>
  )
};

export default ErrorScreen