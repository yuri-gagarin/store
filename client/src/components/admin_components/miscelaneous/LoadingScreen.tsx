import React from 'react'
import { Dimmer, Loader, Segment } from 'semantic-ui-react'

const LoaderExampleText: React.FC<{}> = (props): JSX.Element => {
  return (
    <div style={{ height: "100vh" }}> 
      <Segment style={{ height: "100vh" }}>
        <Dimmer active inverted>
          <Loader>Please Wait</Loader>
        </Dimmer>
      </Segment>
    </div>
  )
};

export default LoaderExampleText