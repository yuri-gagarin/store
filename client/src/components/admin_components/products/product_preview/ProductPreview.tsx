import React from "react";
import { Button, Item } from "semantic-ui-react";
// css //
import "./css/productPreview.css";
// helpers //
import { ConvertDate, setDefaultImage, trimString } from "../../../helpers/displayHelpers";

interface Props {
  product: IProductData;
}

const ProductPreview: React.FC<Props> = ({ product }): JSX.Element => {
  const { name, description, price, images, createdAt } = product;

  return (
    <Item className="productPreviewItem">
      <Item.Image size="small" src={setDefaultImage(images)}/>
      <Item.Content>
        <Item.Meta>
          Displayed name: <strong>{name}</strong>
        </Item.Meta>
        <Item.Meta>
          <span className='productPrice'>Displayed Price: <strong>{price}</strong></span>
          <span className='productCreated'>Created At: <strong>{ConvertDate.international(createdAt)}</strong></span>
        </Item.Meta>
        <Item.Description>{trimString(description, 200)}</Item.Description>
        <Button basic color="green" content="View"></Button>
      </Item.Content>
    </Item>
  );
};

export default ProductPreview;