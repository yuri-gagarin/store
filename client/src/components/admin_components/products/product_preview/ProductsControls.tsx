import React from "react";
import { } from "semantic-ui-react";
// css imports //
import "./css/productsControls.css";
// additional components //

interface Props {
  totalProducts: number;
}
const ProductsControls: React.FC<Props> = ({ totalProducts }): JSX.Element => {

  return (
    <div className="adminProductControlsHolder">
      <div className="adminProductControlsTitle">Product Controls</div>
      <div className="adminProductControls">
        <div className="adminProductCounter">
          <div>Total Products:</div>

          <div className="adminProductsCount">{totalProducts}</div>
        </div>
      </div>
    </div>
  );
};

export default ProductsControls;