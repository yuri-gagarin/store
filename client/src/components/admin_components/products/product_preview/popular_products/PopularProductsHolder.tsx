import React from "react";
import { Item } from "semantic-ui-react";
// additional componets //
import PopularProduct from "./PopularProduct";
// css imports //
import "./css/popularProductsHolder.css";

interface Props {
  popularProducts: IProductData[];
}

const PopularProductHolder: React.FC<Props> = ({ popularProducts }): JSX.Element => {
  return (
    <div className="popProductsHolder">
      <div className="popProductsHolTitle">Popular Products</div>
      <Item.Group divided>
        {
          popularProducts.map((product) => {
            return (
              <PopularProduct key={product._id}  popularProduct={product} />
            );
          })
        }
      </Item.Group>
    </div>  
  );
};

export default PopularProductHolder;