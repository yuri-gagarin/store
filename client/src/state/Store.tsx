import React, { createContext, useReducer } from "react";
import { indexReducer, rootState } from "./reducers/indexReducer";
import { initialStoreState, StoreAction, IStoreState } from "./reducers/storeReducer";
// global app state //
export type IGlobalAppState = {
  storeState: IStoreState;
}
// app actions //
export type AppAction = StoreAction;
// global app context //
export interface IGlobalAppContext {
  state: IGlobalAppState;
  dispatch: React.Dispatch<AppAction>;
}
export type CombinedReducer = (state: IGlobalAppState, action: AppAction) => IGlobalAppState;
// context initialization //
const initialContext: IGlobalAppContext = {
  state: {
    storeState: { ...initialStoreState }
  },
  dispatch: (value: AppAction): void => {}
} 
const Store = createContext<IGlobalAppContext>(initialContext);

export const StateProvider: React.FC<{}> = ({ children }): JSX.Element => {
  const [ globalState, dispatch ] = useReducer<CombinedReducer>(indexReducer, rootState as IGlobalAppState);
  return  (
    <Store.Provider value={{ state: globalState, dispatch: dispatch }}>
      {children}
    </Store.Provider>
  );
};

