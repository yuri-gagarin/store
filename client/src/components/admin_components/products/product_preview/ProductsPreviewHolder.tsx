import React, { useEffect, useState } from "react";
import { Grid, Item } from "semantic-ui-react";
// css imports //
import "./css/productsPreviewHolder.css";
// additional components //
import ProductPreview from "./ProductPreview";
import ProductsControls from "./ProductsControls";
import PopularProductsHolder from "./popular_products/PopularProductsHolder";
import LoadingScreen from "../../miscelaneous/LoadingScreen";
// types and interfaces //
import { AppAction, IGlobalAppState } from "../../../../state/Store";
// api actions //
import { getAllProducts } from "../actions/APIProductActions";

interface Props {
  state: IGlobalAppState;
  dispatch: React.Dispatch<AppAction>;
}

const ProductsPreviewHolder: React.FC<Props> = ({ state, dispatch }): JSX.Element => {
  const { loadedProducts } = state.productState;
  // local state //
  const [ dataLoaded, setDataLoaded ] = useState<boolean>(false);

  useEffect(() => {
    getAllProducts(dispatch)
      .then((success) => {
        if (success) {
          setDataLoaded(true);
        }
      })
  }, []);

  return (
    dataLoaded ? 
    <Grid stackable padded columns={2}>
      <Grid.Row>
      <Grid.Column computer={10} tablet={8} mobile={16}>
        <Item.Group>
          {
            loadedProducts.map((product) => {
              return (
                <ProductPreview 
                  key={product._id}
                  product={product}
                />
              );
            })
          }
        </Item.Group>
      </Grid.Column>
      <Grid.Column computer={6} tablet={8} mobile={16}>
        <ProductsControls totalProducts={loadedProducts.length} />
        <PopularProductsHolder popularProducts={loadedProducts}/>
      </Grid.Column>

      </Grid.Row>
      
    </Grid>
    :
    <LoadingScreen />
  )
};

export default ProductsPreviewHolder;