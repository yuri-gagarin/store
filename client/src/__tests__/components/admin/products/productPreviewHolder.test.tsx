import React from "react";
import { Grid } from "semantic-ui-react"
import { MemoryRouter as Router } from "react-router-dom";
// test dependencies //
import moxios from "moxios";
import { mount, ReactWrapper } from "enzyme"; 
import { act } from "react-dom/test-utils";
// components //
import ProductsPreviewHolder from "../../../../components/admin_components/products/product_preview/ProductsPreviewHolder";
import ProductPreview from "../../../../components/admin_components/products/product_preview/ProductPreview";
import LoadingScreen from "../../../../components/admin_components/miscelaneous/LoadingScreen";
import ErrorScreen from "../../../../components/admin_components/miscelaneous/ErrorScreen";
// state //
import { TestStateProvider } from "../../../../state/Store";
// helpers //
import { generateCleanState } from "../../../../test_helpers/miscHelpers";
import ProductsControls from "../../../../components/admin_components/products/product_preview/ProductsControls";
import PopularProductsHolder from "../../../../components/admin_components/products/product_preview/popular_products/PopularProductsHolder";

describe("ProductPreviewHolder Component render tests", () => {
  let products: IProductData[];

  beforeAll(() => {
    products = [
      {
        _id: "1",
        name: "name",
        price: "100",
        description: "description",
        details: "details",
        images: [],
        createdAt: "now"
      },
      {
        _id: "2",
        name: "name",
        price: "100",
        description: "description",
        details: "details",
        images: [],
        createdAt: "now"
      }
    ];
  });
  // TEST ProductsPreviewHolder in its loading state //
  describe("ProductPreviewHolder in 'loading' state", () => {
    let wrapper: ReactWrapper;

    beforeAll( async () => {
      moxios.install();
      wrapper = mount(
        <Router keyLength={0} initialEntries={["/admin/home/my_products/view_all"]}>
          <TestStateProvider>
              <ProductsPreviewHolder />
          </TestStateProvider>
        </Router>
      );

      await act( async () => {
        await moxios.stubRequest("/api/products", {
          status: 200,
          response: {
            responseMsg: "All Ok",
            products: []
          }
        });
      });
    });
    afterAll(() => {
      moxios.uninstall();
    });
    it("Should correctly render", () => {
      expect(wrapper.find(ProductsPreviewHolder)).toMatchSnapshot();
    });
    it("Should have a LoadingScreen while 'loading == true'", () => {
      const loadingScreen = wrapper.find(LoadingScreen);
      expect(loadingScreen.length).toEqual(1);
    });
    it("Should not display the '#adminProductPreviewHolder'", () => {
      const adminProductPreview = wrapper.find(Grid);
      expect(adminProductPreview.length).toEqual(0);
    }); 
    it("Should not display any ProductPreview Components", () => {
      const productPreviewComponents = wrapper.find(ProductPreview);
      expect(productPreviewComponents.length).toEqual(0);
    });

  });
  // END TEST ProductPreviewHolder in its loading state //
  // TEST ProductPreviewHolder in its loaded state //
  describe("ProductPreview in 'loaded' state", () => {
    let wrapper: ReactWrapper;

    beforeAll( async () => {
      moxios.install();
      const mockState = generateCleanState();

      wrapper = mount(
         <TestStateProvider mockState={mockState}>
           <Router keyLength={0} initialEntries={["/admin/home/my_products/view_all"]}>
             <ProductsPreviewHolder />
           </Router>
         </TestStateProvider>
      );

      await act( async () => {
        await moxios.stubRequest("/api/products", {
          status: 200,
          response: {
            responseMsg: "All ok",
            products: products
          }
        });
      });
      moxios.uninstall();
    });
    
    it("Should correctly render initial 'LoadingScreen' component", () => {
      const loadingScreen = wrapper.find(LoadingScreen);
      const errorScreen = wrapper.find(ErrorScreen);
      const productsGrid = wrapper.find(Grid);
      // assert correct rendering //
      expect(loadingScreen.length).toEqual(1)
      expect(errorScreen.length).toEqual(0);
      expect(productsGrid.length).toEqual(0);
    });
    it("Should correctly render the 'Service' 'Grid' after successful API call", () => {
      wrapper.update();
      const loadingScreen = wrapper.find(LoadingScreen);
      const errorScreen = wrapper.find(ErrorScreen);
      const productsGrid = wrapper.find(Grid);
      // assert correct rendering ///
      expect(loadingScreen.length).toEqual(0);
      expect(errorScreen.length).toEqual(0);
      expect(productsGrid.length).toEqual(1);
      expect(wrapper.find(ProductsPreviewHolder)).toMatchSnapshot();

    });
    it(`Should display a correct number of ProductPreview Components`, () => {
      const productPreviewComponents = wrapper.find(ProductPreview);
      expect(productPreviewComponents.length).toEqual(products.length);
    });
  });
  // END TEST ProductPreviewHolder in its loaded state //
  // TEST ProductPreviewHolder in Error state //
  describe("ProductPreview Component in Error state", () => {
    let wrapper: ReactWrapper; 
    const error = new Error("Error occured");

    beforeAll( async () => {
      const mockState = generateCleanState();

      await act( async () => {
        moxios.install();
        wrapper = mount(
          <Router keyLength={0} initialEntries={["/admin/home/my_products/view_all"]}>
            <TestStateProvider mockState={mockState}>
              <ProductsPreviewHolder />
            </TestStateProvider>
          </Router>
        );
        moxios.stubRequest("/api/products", {
          status: 500,
          response: {
            responseMsg: "Error",
            error: error
          }
        });
      });
      moxios.uninstall();
    });
    afterAll(() => {
      moxios.uninstall();
    });
    it("Should render the initial Loading Screen after an  API call", () => {
      // wrapper.update();
      const loadingScreen = wrapper.find(LoadingScreen);
      const errorScreen = wrapper.find(ErrorScreen);
      const productGrid = wrapper.find(Grid);
      // assert correct rendering //
      expect(loadingScreen.length).toEqual(1);
      expect(errorScreen.length).toEqual(0);
      expect(productGrid.length).toEqual(0);
    });
    it("Should render the 'ErrorScreen' Component only", () => {
      wrapper.update();
      const loadingScreen = wrapper.find(LoadingScreen);
      const errorScreen = wrapper.find(ErrorScreen);
      const productGrid = wrapper.find(Grid);
      // assert correct rendering //
      expect(errorScreen.length).toEqual(1);
      expect(loadingScreen.length).toEqual(0);
      expect(productGrid.length).toEqual(0)
    });
  
    it("Should properly redispatch last API call from 'ErrorScreen' Component", async () => {
      await act( async () => {
        moxios.install();
        moxios.stubRequest("/api/products", {
          status: 200,
          response: {
            responseMsg: "All Ok",
            products: products
          }
        });
        
        const retryButton = wrapper.find("#errorScreenRetryButton");
        retryButton.at(0).simulate("click");
      });
      // assert difference //
      expect(wrapper.find(ErrorScreen).length).toEqual(1);
    });
    it("Should NOT render the 'LoadingScreen' Component", () => {
      wrapper.update();
      // assert new render //
      const loadingScreen = wrapper.find(LoadingScreen);
      expect(loadingScreen.length).toEqual(0);
    })
    it("Should render the 'ProductPreviewHolder' Grid", () => {
      const previewGrid = wrapper.find(ProductsPreviewHolder).find(Grid);
      expect(previewGrid.length).toEqual(1);
    });
    it(`Should display a correct number of ProductPreview Components`, () => {
      const productPreviewComponents = wrapper.find(ProductPreview);
      expect(productPreviewComponents.length).toEqual(2);
    });
    it("Should render 'ProductsControls' Component", () => {
      const productsControls = wrapper.find(ProductsControls);
      expect(productsControls.length).toEqual(1);
    });
    it("Should render 'PopularProductsHolder' Component", () => {
      const popularProductsHolder = wrapper.find(PopularProductsHolder);
      expect(popularProductsHolder.length).toEqual(1);
    });
    
  });
  
  // END TEST ProductsPreviewHolder test with API error returned //
});