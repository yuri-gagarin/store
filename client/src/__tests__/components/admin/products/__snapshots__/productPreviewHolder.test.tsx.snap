// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductPreviewHolder Component render tests 'ProductPreviewHolder' after a successful API call Should correctly render the 'Service' 'Grid' after successful API call 1`] = `
<withRouter(ProductsPreviewHolder)>
  <ProductsPreviewHolder
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "pathname": "/admin/home/my_products/view_all",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/admin/home/my_products/view_all",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
    location={
      Object {
        "hash": "",
        "pathname": "/admin/home/my_products/view_all",
        "search": "",
        "state": undefined,
      }
    }
    match={
      Object {
        "isExact": false,
        "params": Object {},
        "path": "/",
        "url": "/",
      }
    }
  >
    <Grid
      columns={2}
      padded={true}
      stackable={true}
    >
      <div
        className="ui stackable padded two column grid"
      >
        <GridRow>
          <div
            className="row"
          >
            <GridColumn
              computer={10}
              mobile={16}
              tablet={8}
            >
              <div
                className="ten wide computer sixteen wide mobile eight wide tablet column"
              >
                <ItemGroup>
                  <div
                    className="ui items"
                  >
                    <ProductPreview
                      key="1"
                      product={
                        Object {
                          "_id": "1",
                          "createdAt": "now",
                          "description": "description",
                          "details": "details",
                          "images": Array [],
                          "name": "name",
                          "price": "100",
                        }
                      }
                    >
                      <Item
                        className="productPreviewItem"
                      >
                        <div
                          className="item productPreviewItem"
                        >
                          <ItemImage
                            size="small"
                            src="https://react.semantic-ui.com/images/wireframe/image.png"
                          >
                            <Image
                              as="img"
                              size="small"
                              src="https://react.semantic-ui.com/images/wireframe/image.png"
                              ui={true}
                              wrapped={true}
                            >
                              <div
                                className="ui small image"
                              >
                                <img
                                  src="https://react.semantic-ui.com/images/wireframe/image.png"
                                />
                              </div>
                            </Image>
                          </ItemImage>
                          <ItemContent>
                            <div
                              className="content"
                            >
                              <ItemMeta>
                                <div
                                  className="meta"
                                >
                                  Displayed name: 
                                  <strong>
                                    name
                                  </strong>
                                </div>
                              </ItemMeta>
                              <ItemMeta>
                                <div
                                  className="meta"
                                >
                                  <span
                                    className="productPrice"
                                  >
                                    Displayed Price: 
                                    <strong>
                                      100
                                    </strong>
                                  </span>
                                  <span
                                    className="productCreated"
                                  >
                                    Created At: 
                                    <strong>
                                      NaN-undefined-NaN
                                    </strong>
                                  </span>
                                </div>
                              </ItemMeta>
                              <ItemDescription>
                                <div
                                  className="description"
                                >
                                  description...
                                </div>
                              </ItemDescription>
                              <Button
                                as="button"
                                basic={true}
                                color="green"
                                content="View"
                              >
                                <Ref
                                  innerRef={
                                    Object {
                                      "current": <button
                                        class="ui green basic button"
                                      >
                                        View
                                      </button>,
                                    }
                                  }
                                >
                                  <RefFindNode
                                    innerRef={
                                      Object {
                                        "current": <button
                                          class="ui green basic button"
                                        >
                                          View
                                        </button>,
                                      }
                                    }
                                  >
                                    <button
                                      className="ui green basic button"
                                      onClick={[Function]}
                                    >
                                      View
                                    </button>
                                  </RefFindNode>
                                </Ref>
                              </Button>
                            </div>
                          </ItemContent>
                        </div>
                      </Item>
                    </ProductPreview>
                    <ProductPreview
                      key="2"
                      product={
                        Object {
                          "_id": "2",
                          "createdAt": "now",
                          "description": "description",
                          "details": "details",
                          "images": Array [],
                          "name": "name",
                          "price": "100",
                        }
                      }
                    >
                      <Item
                        className="productPreviewItem"
                      >
                        <div
                          className="item productPreviewItem"
                        >
                          <ItemImage
                            size="small"
                            src="https://react.semantic-ui.com/images/wireframe/image.png"
                          >
                            <Image
                              as="img"
                              size="small"
                              src="https://react.semantic-ui.com/images/wireframe/image.png"
                              ui={true}
                              wrapped={true}
                            >
                              <div
                                className="ui small image"
                              >
                                <img
                                  src="https://react.semantic-ui.com/images/wireframe/image.png"
                                />
                              </div>
                            </Image>
                          </ItemImage>
                          <ItemContent>
                            <div
                              className="content"
                            >
                              <ItemMeta>
                                <div
                                  className="meta"
                                >
                                  Displayed name: 
                                  <strong>
                                    name
                                  </strong>
                                </div>
                              </ItemMeta>
                              <ItemMeta>
                                <div
                                  className="meta"
                                >
                                  <span
                                    className="productPrice"
                                  >
                                    Displayed Price: 
                                    <strong>
                                      100
                                    </strong>
                                  </span>
                                  <span
                                    className="productCreated"
                                  >
                                    Created At: 
                                    <strong>
                                      NaN-undefined-NaN
                                    </strong>
                                  </span>
                                </div>
                              </ItemMeta>
                              <ItemDescription>
                                <div
                                  className="description"
                                >
                                  description...
                                </div>
                              </ItemDescription>
                              <Button
                                as="button"
                                basic={true}
                                color="green"
                                content="View"
                              >
                                <Ref
                                  innerRef={
                                    Object {
                                      "current": <button
                                        class="ui green basic button"
                                      >
                                        View
                                      </button>,
                                    }
                                  }
                                >
                                  <RefFindNode
                                    innerRef={
                                      Object {
                                        "current": <button
                                          class="ui green basic button"
                                        >
                                          View
                                        </button>,
                                      }
                                    }
                                  >
                                    <button
                                      className="ui green basic button"
                                      onClick={[Function]}
                                    >
                                      View
                                    </button>
                                  </RefFindNode>
                                </Ref>
                              </Button>
                            </div>
                          </ItemContent>
                        </div>
                      </Item>
                    </ProductPreview>
                  </div>
                </ItemGroup>
              </div>
            </GridColumn>
            <GridColumn
              computer={6}
              mobile={16}
              tablet={8}
            >
              <div
                className="six wide computer sixteen wide mobile eight wide tablet column"
              >
                <ProductsControls
                  totalProducts={2}
                >
                  <div
                    className="adminProductControlsHolder"
                  >
                    <div
                      className="adminProductControlsTitle"
                    >
                      Product Controls
                    </div>
                    <div
                      className="adminProductControls"
                    >
                      <div
                        className="adminProductCounter"
                      >
                        <div>
                          Total Products:
                        </div>
                        <div
                          className="adminProductsCount"
                        >
                          2
                        </div>
                      </div>
                    </div>
                  </div>
                </ProductsControls>
                <PopularProductHolder
                  popularProducts={
                    Array [
                      Object {
                        "_id": "1",
                        "createdAt": "now",
                        "description": "description",
                        "details": "details",
                        "images": Array [],
                        "name": "name",
                        "price": "100",
                      },
                      Object {
                        "_id": "2",
                        "createdAt": "now",
                        "description": "description",
                        "details": "details",
                        "images": Array [],
                        "name": "name",
                        "price": "100",
                      },
                    ]
                  }
                >
                  <div
                    className="popProductsHolder"
                  >
                    <div
                      className="popProductsHolTitle"
                    >
                      Popular Products
                    </div>
                    <ItemGroup
                      divided={true}
                    >
                      <div
                        className="ui divided items"
                      >
                        <PopularProduct
                          key="1"
                          popularProduct={
                            Object {
                              "_id": "1",
                              "createdAt": "now",
                              "description": "description",
                              "details": "details",
                              "images": Array [],
                              "name": "name",
                              "price": "100",
                            }
                          }
                        >
                          <Item>
                            <div
                              className="item"
                            >
                              <ItemImage
                                size="tiny"
                                src="https://react.semantic-ui.com/images/wireframe/image.png"
                              >
                                <Image
                                  as="img"
                                  size="tiny"
                                  src="https://react.semantic-ui.com/images/wireframe/image.png"
                                  ui={true}
                                  wrapped={true}
                                >
                                  <div
                                    className="ui tiny image"
                                  >
                                    <img
                                      src="https://react.semantic-ui.com/images/wireframe/image.png"
                                    />
                                  </div>
                                </Image>
                              </ItemImage>
                              <ItemContent>
                                <div
                                  className="content"
                                >
                                  <ItemHeader>
                                    <div
                                      className="header"
                                    >
                                      name
                                    </div>
                                  </ItemHeader>
                                  <ItemMeta>
                                    <div
                                      className="meta"
                                    >
                                      <span>
                                        Priced At: 
                                        <strong>
                                          100
                                        </strong>
                                      </span>
                                      <span>
                                        -
                                      </span>
                                      <span>
                                        Posted: 
                                        <strong>
                                          NaN-undefined-NaN
                                        </strong>
                                      </span>
                                    </div>
                                  </ItemMeta>
                                  <Button
                                    as="button"
                                    basic={true}
                                    color="blue"
                                    content="More..."
                                  >
                                    <Ref
                                      innerRef={
                                        Object {
                                          "current": <button
                                            class="ui blue basic button"
                                          >
                                            More...
                                          </button>,
                                        }
                                      }
                                    >
                                      <RefFindNode
                                        innerRef={
                                          Object {
                                            "current": <button
                                              class="ui blue basic button"
                                            >
                                              More...
                                            </button>,
                                          }
                                        }
                                      >
                                        <button
                                          className="ui blue basic button"
                                          onClick={[Function]}
                                        >
                                          More...
                                        </button>
                                      </RefFindNode>
                                    </Ref>
                                  </Button>
                                </div>
                              </ItemContent>
                            </div>
                          </Item>
                        </PopularProduct>
                        <PopularProduct
                          key="2"
                          popularProduct={
                            Object {
                              "_id": "2",
                              "createdAt": "now",
                              "description": "description",
                              "details": "details",
                              "images": Array [],
                              "name": "name",
                              "price": "100",
                            }
                          }
                        >
                          <Item>
                            <div
                              className="item"
                            >
                              <ItemImage
                                size="tiny"
                                src="https://react.semantic-ui.com/images/wireframe/image.png"
                              >
                                <Image
                                  as="img"
                                  size="tiny"
                                  src="https://react.semantic-ui.com/images/wireframe/image.png"
                                  ui={true}
                                  wrapped={true}
                                >
                                  <div
                                    className="ui tiny image"
                                  >
                                    <img
                                      src="https://react.semantic-ui.com/images/wireframe/image.png"
                                    />
                                  </div>
                                </Image>
                              </ItemImage>
                              <ItemContent>
                                <div
                                  className="content"
                                >
                                  <ItemHeader>
                                    <div
                                      className="header"
                                    >
                                      name
                                    </div>
                                  </ItemHeader>
                                  <ItemMeta>
                                    <div
                                      className="meta"
                                    >
                                      <span>
                                        Priced At: 
                                        <strong>
                                          100
                                        </strong>
                                      </span>
                                      <span>
                                        -
                                      </span>
                                      <span>
                                        Posted: 
                                        <strong>
                                          NaN-undefined-NaN
                                        </strong>
                                      </span>
                                    </div>
                                  </ItemMeta>
                                  <Button
                                    as="button"
                                    basic={true}
                                    color="blue"
                                    content="More..."
                                  >
                                    <Ref
                                      innerRef={
                                        Object {
                                          "current": <button
                                            class="ui blue basic button"
                                          >
                                            More...
                                          </button>,
                                        }
                                      }
                                    >
                                      <RefFindNode
                                        innerRef={
                                          Object {
                                            "current": <button
                                              class="ui blue basic button"
                                            >
                                              More...
                                            </button>,
                                          }
                                        }
                                      >
                                        <button
                                          className="ui blue basic button"
                                          onClick={[Function]}
                                        >
                                          More...
                                        </button>
                                      </RefFindNode>
                                    </Ref>
                                  </Button>
                                </div>
                              </ItemContent>
                            </div>
                          </Item>
                        </PopularProduct>
                      </div>
                    </ItemGroup>
                  </div>
                </PopularProductHolder>
              </div>
            </GridColumn>
          </div>
        </GridRow>
      </div>
    </Grid>
  </ProductsPreviewHolder>
</withRouter(ProductsPreviewHolder)>
`;
